Version 5: ACID-Transactions and Typescript

Uzlopak opened the pull request:
@koresar 
I open this PR, so that others can also do a code review :)

clintonb reviewed (changes requested):
Can this be split into multiple pull requests? This is _way_ too big to merge without taking on significant risks.

Uzlopak commented:
@clintonb 

No this is work of 5 days and night work. I will not rip this apart. 

Also despite that it shows that 45 files were changed, only only 13 are active code files. The rest are configuration files, unit tests, interfaces etc.. So if you only focus on the 13 files in the src folder, you can manage it. Also keep in mind, that you have 100% test coverage. You can e.g. go to book.spec.ts, make `.only` for the book-describe, than e.g. select `balance` and do a `npm run test:coverage` and check if the code is covered by 100% or not. If 100% than check the unit tests if they are really complete and do make sense or not. Or start with the helper methods. 

Please clone the repo and read the code with your favorite IDE. This would already help alot. :)

Uzlopak commented on src/Book.ts:
```diff
@@ -0,0 +1,197 @@
+  constructor(name: string, options = {} as { precision?: number }) {
+    this.name = name;
+    this.precision =
+      typeof options.precision === "number" ? options.precision : 7;
```
Maybe should also add a Number.isInteger check. Maybe makes sense, to throw an Error, as it could have reliability implications.E.g. somebody sets precision to "8" (<--- string) and expects that it will work but it would not have any impact and maybe loses some satoshis. 

Any thoughts?

Uzlopak replied:
Added validation.

Uzlopak commented on src/Book.ts:
```diff
@@ -0,0 +1,197 @@
+    let skip: PipelineStage.Skip | undefined = undefined;
+
+    if (query.perPage) {
+      skip = { $skip: (query.page ? query.page - 1 : 0) * query.perPage };
```
Balance only uses skip for the pagination but not a limit. This results in the effect, that the balance sums up. E.g. you do a .ledger call with the same parameters and balance and ledger will not sum to the same value. 

A potential fix would be https://github.com/flash-oss/medici/commit/50d928f90bd2b518e51deee75452c1851d939ca9

Maybe pagination in balance makes no sense anyway, and should be removed.

To everybody:
What do you think?

Uzlopak commented on src/models/journals.ts:
```diff
@@ -0,0 +1,189 @@
+
+  for (const trans of transactions) {
+    const meta: IAnyObject = {};
+    Object.keys(trans.toObject()).forEach((key) => {
```
I think the whole conversion makes no sense in 99.9999999999999% of all cases.

The only case were I could think of a potential use case, was when you for some reason migrate entries on the fly. But why would you do this anyway?

This could be a simple const meta = trans.meta;

I did not remove this code, as the performance is not touched by these lines as the slowest part is actually the mongoDB and not this code.

But theoretically removing this, would simplify the code.

Uzlopak replied:
Meh, not a performance issue imho. Better not touch this for now.

Uzlopak commented on src/models/transactions.ts:
```diff
@@ -0,0 +1,100 @@
+  });
+  schema.index({ "account_path.0": 1, book: 1, approved: 1 });
+  schema.index({
+    "account_path.0": 1,
```
Usually the indizes should be sorted by cardinality of each item. See https://docs.mongodb.com/manual/core/index-compound/#prefixes

If we think, that Accounting are structured like 

"Expenses:Salary:Employer1"
"Expenses:Salary:Employer2"
"Expenses:Salary:Employer3"

Than the indizes should be in the reverse order, like `account_path.1` should be before `account_path.0` as the cardinality would be in the reverse order. But if you have Accounts structured like

"Employer1:Salary:Expenses"
"Employer2:Salary:Expenses"
"Employer3:Salary:Expenses"

than current indizes make sense.

But somehow I think it is actually the first case, as the README.md suggests that structure. Maybe we should determine what the most used account naming style is, and then recommend it as best practice. Maybe we should also add a syncIndices helper method as with setTransactionSchema you could add your own indizes and after you set the transactionSchema with setTransactionSchema you could use the syncIndizes-helper to enforce the indizes.

Maybe setTransactionSchema should have an Option parameter, were you can select if you want to add the default-indizes or not.

to Everyone:
Any thoughts?

Uzlopak replied:
Added syncIndexes.

Uzlopak replied:
Adding an index datetime and timestamp improves the performance of ledger noticable.

Uzlopak replied:
Added documentation about cardinality.

Uzlopak commented on src/models/journals.ts:
```diff
@@ -0,0 +1,189 @@
+    )
+    .exec();
+
+  for (let i = 0, il = transactions.length; i < il; i++) {
```
I think, we could safely do this in a updateMany operation, as voided and void_reason is not used for the newly created transactions. 

Any thoughts?

Uzlopak replied:
No, if somebody adds a pre('save') to his custom schema, then an updateMany would not use the middleware.

Uzlopak commented on src/helper/parseQuery.ts:
```diff
@@ -0,0 +1,70 @@
+    const key = keys[i];
+    const value = extra[key];
+    filterQuery[isValidTransactionKey(key) ? key : `meta.${key}`] =
+      typeof value === "string" &&
```
We could theoretically use the mongoose Schema of the transactionModel to determine if the key is actually an ObjectId removing the undocumented behaviour that a field has to start with `_` or and with `_id` to make an implicit conversion to an ObjectId.

Uzlopak replied:
Adding a separate index on datetime and timestamp improves pagination of ledger().

Uzlopak replied:
implemented extractObjectIdKeysFromSchema

Uzlopak commented on src/helper/parseAccountField.ts:
```diff
@@ -0,0 +1,31 @@
+  const filterQuery: FilterQuery<ITransaction> = {};
+  let i, il, j, jl;
+
+  if (Array.isArray(account) && account.length === 1) {
```
This maybe has a low perf impact but still want to mention:
We could theoretically determine if the account is a single string and if it contains three parts. If it contains three parts, we could just search for accounts: account.

Uzlopak replied:
adapted. It is 6 times faster...

Uzlopak commented on src/TransactionError.ts:
```diff
@@ -0,0 +1,10 @@
+export class TransactionError extends Error {
+  public code = 400;
+  public total: number;
```
total is not really descriptive...

Uzlopak commented on src/Entry.ts:
```diff
@@ -0,0 +1,182 @@
+      account_path = account_path.split(":");
+    }
+
+    if (account_path.length > 3) {
```
we could actually make this configurable.

Uzlopak replied:
made it configurable and is used by parseQuery.

clintonb commented:
It’s nice you’ve done this work, but I’m not reviewing it. This isn’t my project, and the scope of the pull request is simply too large. 

You already have separate commits. I _highly_ recommend breaking this into multiple pull requests to make the work easier to review. As a maintainer of other open source projects, it’s far easier/faster for me to review small, focused, pull requests than giant ones like this. 100% test coverage does not mean 100% bug-free.

Uzlopak commented on src/Book.ts:
```diff
@@ -0,0 +1,197 @@
+    reason?: undefined | string,
+    options = {} as IOptions
+  ) {
+    const journal: TJournalDocument<J> = (await journalModel
```
Maybe we should throw an Error if we dont find the requested Journal?

Uzlopak replied:
Now it is book-scoped and throws an error if we dont find thejournal within the book

Uzlopak commented on src/Book.ts:
```diff
@@ -0,0 +1,197 @@
+    const q = transactionModel
+      .find(filterQuery, undefined, options)
+      .lean(lean)
+      .sort({
```
We always sort by datetime and timestamp. ParseQuery always breaks our accounts to account_path Array. This makes the index
  schema.index({
    accounts: 1,
    book: 1,
    approved: 1,
    datetime: -1,
    timestamp: -1,
  });

basically useless. 

Also we basically never enforce approved, despite the fact that boolean fields should be the last field in the index. I guess this index is never really used.

I think all the account_path indizes should have datetime and timestamp and the approved flag should be removed or be the last element of the compound index. This could significantly improve the performance of balance and ledger.

Uzlopak replied:
Added maxAmountPath and modified parseQuery. Now the index can be used.

Uzlopak replied:
Wow. It is 6x faster to use that index.

Uzlopak commented:
@clintonb 
I know what you mean. In OSS you have alot of Contributors and it is easier to review one small PR than a whole rewrite. But my experience as a professional Software Developery is that you have sometimes get out of the comfort zone and do the hard work and research to get the best result. :) Also this package is fairly simple. 

And breaking it to multiple PRs is not an option, so I have to decline your suggestion. I mean, come on... these are basically 13 files. I have regularly PRs with more than hundred file changes... Not ideal, but you get used to it. And having a good code coverage and a good test structure helps you to review each file in a fair amount of time. If you of course do a unit test which does everything, than yeah you get 100% coverage fairly easy. But that garbage gets rejected by me also regularly. 

Best Regards :)

Uzlopak commented:
@koresar 

Good morning to Australia. Medici should have now an awesome status. Only the pagination in balance is kind of inconsistent. Maybe you can tell if and what I should do with it? Keep it as is? Rip out Pagination? Fix Pagination?

Other than that, my PR is so far ready. I dont know what else you can do to improve the performance.

Uzlopak commented:
The current Changelog


  - The project was rewritten with TypeScript. Types are provided within the package now.
  - Added support for MongoDB sessions (aka ACID transactions). See `IOptions` type.
  - Added a `mongoTransaction`-method, which is a convenience shortcut for `mongoose.connection.transaction`.
  - Added async helper method `initModels`, which initializes the underlying `transactionModel` and `journalModel`. Use this after you connected to  the MongoDB-Server if you want to use transactions. Or else you could get `Unable to read from a snapshot due to pending collection catalog changes; please retry the operation.`-Error when acquiring a session because the actual database-collection is still being created by the underlying mongoose-instance.
  - Node.js 12 is the lowest supported version. Although, 10 should still work fine, when using mongoose v5.
  - Mongoose v6 is the only supported version now. Avoid using both v5 and v6 in the same project.
  - MongoDB 4 and above is supported. You can still use MongoDB 3, but ACID-sessions could have issues.
  - You can't import `book` anymore. Only `Book` is supported. `require("medici").Book`.
  - Added a new index on the transactionModel to improve paginated ledger queries.
  - `.ledger()` returns lean Transaction-Objects for better performance. To retrieve hydrated Transaction-Objects, set lean to false in the third parameter of `.ledger()`. It is recommended to not hydrate the transactions, as it implies that the transactions could be manipulated and the data integrity of Medici could be risked. 
  - You can now specify the `precision`. Book now accepts an optional second parameter, where you can set the `precision` used internally by Medici. Default value is 7 digits precision. Javascript has issues with floating points precision and can only handle 16 digits precision, like 0.1 + 0.2 results in 0.30000000000000004 and not 0.3. The default precision of 7 digits after decimal, results in the correct result of 0.1 + 0.2 = 0.3. The default value is taken from medici version 4.0.2. Be careful, if you use currency, which has more decimal points, e.g. Bitcoin has a precision of 8 digits after the comma. So for Bitcoin you should set the precision to 8. You can enforce an "only-Integer"-mode, by setting the precision to 0. But keep in mind that Javascript has a max safe integer limit of 9007199254740991.
  - Added `maxAccountPath`. You can set the maximum amount of account paths via the second parameter of Book. This can improve the performance of `.balance()` and `.ledger()` calls as it will then use the accounts attribute of the transactions as a filter.
  - Added validation for `name` of Book, `maxAccountPath` and `precision`. A `name` has to be not an empty string or a string containing only whitespaces. `precision` has to be an integer bigger or equal 0. `maxAccountPath` has to be an integer bigger or equal 0. 
  - Added `setJournalSchema` and `setTransactionSchema` to use custom Schemas. It will ensure, that all relevant middlewares and methods are also added when using custom Schemas. Use `syncIndexes`-method from medici after setTransactionSchema to enforce the defined indexes on the models.
  - Added prototype-pollution protection when creating entries. Reserved words like `__proto__` can not be used as properties of a Transaction or a Journal or their meta-Field as they will get silently filtered.
  - When calling `book.void()` the provided `journal_id` has to belong to the `book`. If the journal does not exist within the book, medici will throw a `JournalNotFoundError`. In medici < 5 you could theoretically void a `journal` of another `book`. 
  - Added a `trackAccountChanges`-model to make it possible to call `.balance()` while using a mongo-session and get.

Uzlopak commented:
@koresar If you want to remove paginated balance #36

Uzlopak commented on src/models/transactions.ts:
```diff
@@ -0,0 +1,163 @@
+
+    await lockModel.collection.updateOne(
+      { account, book },
+      { $inc: { __v: 1 } },
```
Why does this work? Should be also containing $setOnInsert: {book, account}

Not that mongo just creates a general document and blocks everything

Uzlopak replied:
```suggestion
      { $setOnInsert: { book, account }, $inc: { __v: 1 } },
```

Uzlopak commented:
@nicolasburtey 
Implemented a lockAccounts method on book, which should be called as last operation in the transaction. This improves the performance of transactions in my unit test drastically. Before it had about 700 ms duration. Now it has about 200 ms duration. I took this idea from MongoDB Performance Tuning (2021), p. 217

```
If we move the contentious statement to the end of the transaction, then the chance
of a TransientTransactionError will be reduced, since the window for conflict will be
reduced to the final few moments in the execution of the transaction.
```

@koresar + @nicolasburtey 

I am not that happy with the name `.lockAccounts()`. Do you have any suggestions? 

But yeah, I guess the performance improvement is not to neglect. :)

Uzlopak commented on src/models/locks.ts:
```diff
@@ -0,0 +1,43 @@
+  {
+    updatedAt: 1,
+  },
+  { expireAfterSeconds: 3600 }
```
```suggestion
  { expireAfterSeconds: 60 * 60 * 24 }
```

Uzlopak commented:
@koresar 
I am gone for 10 days. I hope the code is good to merge. Looking forward

koresar commented:
Hi. I'm going through changes commit by commit. I'm loving the fact you are doing small commits rather than pushing huge ball of code in one go!

Random thoughts in no order:

* We can make the default precision as 8. It won't break anything for anybody if me move from 7 to 8. (Trust me, I did it couple of times by now.)

* Instead of `for` loop I'd recommend using `for-of`. It is as fast as `for` loop nowadays AFAIK.

```ts
export function extractObjectIdKeysFromSchema(schema: Schema) {
  const result: Set<string> = new Set();
  for (const [key, value] of Object.entries(schema.paths)) {
    if (value instanceof Schema.Types.ObjectId) {
      result.add(key);
    }
  }
  return result;
}
```

* I am loving all the features, fixed, hardening, etc you did. The index cardinality got me excited!

* Some of the changes labelled as "BREAKING" in README are more like "POTENTIALLY BREAKING". Meaning that it would not break most people code. E.g. void hardening, proto pollution protection, etc.

IMPORTANT!

The `typescript-migration` branch contains lock "account" feature code. That's no good. This PR is too much to merge.
Could you please contact me in twitter DM?

Uzlopak commented:
Regarding the for loop: 
I prefer the Classic for loops as they were always the fastest. Actually using for(var i = 0, il = arr.length ... Is the fastest loop and not using let. Also with the new stuff it could be that the memory consumption is bigger, as it potentially creates new Objects and Variables. But I guess we could easily argue that not the for loop is the bottleneck but the database operations.

So I am totally Ok if you want to use for-of instead of classical for-loo

koresar commented:
I'd change it to `for-of` because code is mostly read than written. We should write for readability.
The current `for` loop usage is premature optimisaiton IMO.

koresar commented:
We'll continue development in the `next` branch. Without any PRs for now.
I'll publish `v5.0.0-next` to npm.
